syspagegen/syspagegen.c

```diff
@@ -0,0 +1,822 @@
+{
+	char *end;
+	unsigned int i;
+	addr_t addr = 0;
+	size_t size = 0, len;
+	struct phfs_alias_t *alias;
+
+	if (argc != 4) {
+		fprintf(stderr, "\n%s: Wrong argument count", argv[0]);
+		return -1;
+	}
+
+	for (i = 2; i < 4; ++i) {
+		if (i == 2)
+			addr = strtoul(argv[i], &end, 0);
+		else if (i == 3)
+			size = strtoul(argv[i], &end, 0);
+
+		if (*end) {
+			fprintf(stderr, "\n%s: Wrong arguments", argv[0]);
+			return -1;
+		}
+	}
+
+	alias = malloc(sizeof(struct phfs_alias_t));
+	if (alias == NULL)
+		return -1;
+
+	len = strlen(argv[1]);
+	len = (len < (sizeof(alias->name) - 1)) ? (len + 1) : (sizeof(alias->name) - 1);
+	memcpy(alias->name, argv[1], len);
+	alias->name[len] = '\0';
+
+	/* Update max image size */
+	if (sysgen_common.syspage->hs.imgsz < (addr + size))
+		sysgen_common.syspage->hs.imgsz = addr + size;
+
+	alias->addr = addr + sysgen_common.pkernel;
+	alias->size = size;
+
+	LIST_INSERT_HEAD(&sysgen_common.aliases, alias, ptrs);
+
+	return 0;
+}
+
+
+static int sysgen_mapOverlapping(const char *name, addr_t start, addr_t end)
+{
+	sysptr_t ptr;
+	const syspage_map_t *map;
+
+	if (sysgen_common.syspage->maps != 0) {
+		ptr = sysgen_common.syspage->maps;
+		do {
+			map = CAST_SYSPTR(syspage_map_t *, ptr);
+			if (((map->start < end) && (map->end > start)) ||
+					(strcmp(name, CAST_SYSPTR(char *, map->name)) == 0))
+				return -1;
+
+		} while ((ptr = map->next) != sysgen_common.syspage->maps);
+	}
+
+	return 0;
+}
+
+
+static int sysgen_mapNameResolve(const char *name, uint8_t *id)
+{
+	sysptr_t ptr;
+	const syspage_map_t *map;
+
+	if (sysgen_common.syspage->maps != 0) {
+		ptr = sysgen_common.syspage->maps;
+		do {
+			map = CAST_SYSPTR(syspage_map_t *, ptr);
+
+			if (strcmp(name, CAST_SYSPTR(char *, map->name)) == 0) {
+				*id = map->id;
+				return 0;
+			}
+
+		} while ((ptr = map->next) != sysgen_common.syspage->maps);
+	}
+
+	return -1;
+}
+
+
+static int sysgen_strAttr2ui(const char *str, unsigned int *attr)
+{
+	int i;
+
+	*attr = 0;
+	for (i = 0; str[i]; ++i) {
+		switch (str[i]) {
+			case 'r':
+				*attr |= mAttrRead;
+				break;
+			case 'w':
+				*attr |= mAttrWrite;
+				break;
+			case 'x':
+				*attr |= mAttrExec;
+				break;
+			case 's':
+				*attr |= mAttrShareable;
+				break;
+			case 'c':
+				*attr |= mAttrCacheable;
+				break;
+			case 'b':
+				*attr |= mAttrBufferable;
+				break;
+			default:
+				fprintf(stderr, "\nsysgen: Wrong attribute - '%c'", str[i]);
+				return -1;
+		}
+	}
+
+	return 0;
+}
+
+
+static int sysgen_mapAdd(const char *mapName, addr_t start, addr_t end, unsigned int attr)
+{
+	size_t len;
+	char *name;
+	sysptr_t ptr;
+	syspage_map_t *map, *headMap;
+
+	ptr = sysgen_buffAlloc(sizeof(syspage_map_t));
+	if (ptr == 0)
+		return -1;
+
+	map = CAST_SYSPTR(syspage_map_t *, ptr);
+
+	len = strlen(mapName);
+	map->name = sysgen_buffAlloc(len + 1);
+	if (map->name == 0)
+		return -1;
+
+	name = CAST_SYSPTR(char *, map->name);
+	memcpy(name, mapName, len);
+	name[len] = '\0';
```

Comment:
We are sure mapName is NULL terminated, no need to place it again
```suggestion
	memcpy(name, mapName, len + 1);
```

Reply:
Done